packages/dashboard/app/withdraw/utils/alchemyUtils.tsx

```diff
@@ -0,0 +1,171 @@
+import * as ethers from 'ethers';
+
+const BASE_RPC_URL = 'https://mainnet.base.org';
+const ALCHEMY_BASE_URL = `https://base-mainnet.g.alchemy.com/v2/${process.env.NEXT_PUBLIC_ALCHEMY_API_KEY}`;
+
+interface TokenBalance {
+  address: string;
+  symbol: string;
+  name: string;
+  balance: string;
+  rawBalance: ethers.BigNumber;
+  decimals: number;
+  logoUrl?: string;
+}
+
+export interface TokenBalanceResult {
+  success: boolean;
+  balances: TokenBalance[];
+  error?: string;
+  errorDetails?: string;
+}
+
+interface AlchemyTokenBalance {
+  contractAddress: string;
+  tokenBalance: string;
+  error?: string;
+}
+
+interface AlchemyTokenMetadata {
+  decimals?: number;
+  logo?: string;
+  name?: string;
+  symbol?: string;
+}
+
+const formatBalance = (balance: ethers.BigNumber, decimals: number): string => {
+  return ethers.utils.formatUnits(balance, decimals);
+};
+
+export const fetchTokenBalances = async (ethAddress: string): Promise<TokenBalanceResult> => {
+  try {
+    const provider = new ethers.providers.JsonRpcProvider(BASE_RPC_URL);
+
+    // Check ETH balance without using Alchemy
+    const ethBalance = await provider.getBalance(ethAddress);
+    const tokenBalances: TokenBalance[] = [{
+      address: 'ETH',
+      symbol: 'ETH',
+      name: 'Ethereum',
+      balance: formatBalance(ethBalance, 18),
+      rawBalance: ethBalance,
+      decimals: 18
+    }];
+
+    // Get all token balances for the address - Specific to Base Mainnet
+    const tokensResponse = await fetch(ALCHEMY_BASE_URL, {
+      method: 'POST',
+      headers: { 'Content-Type': 'application/json' },
+      body: JSON.stringify({
+        jsonrpc: '2.0',
+        id: 1,
+        method: 'alchemy_getTokenBalances',
+        params: [ethAddress]
+      })
+    });
+
+    if (!tokensResponse.ok) {
+      return {
+        success: false,
+        balances: tokenBalances, // Return ETH balance at least
+        error: 'Failed to fetch token balances from Alchemy',
+        errorDetails: `Status: ${tokensResponse.status}, StatusText: ${tokensResponse.statusText}`
+      };
+    }
+
+    const tokensData = await tokensResponse.json();
+    
+    if (tokensData.error) {
+      return {
+        success: false,
+        balances: tokenBalances, // Return ETH balance at least
+        error: 'Alchemy API returned an error',
+        errorDetails: JSON.stringify(tokensData.error)
+      };
+    }
+    
+    const tokenBalancesList = tokensData.result?.tokenBalances || [];
+    const nonZeroBalances = tokenBalancesList.filter((token: AlchemyTokenBalance) => {
+      if (token.error) return false;
+      const balanceBN = ethers.BigNumber.from(token.tokenBalance);
+      return !balanceBN.isZero();
+    });
+
+    // If we have any non-zero balances, we need to get the token metadata
+    if (nonZeroBalances.length > 0) {
+      const tokenDetailsPromises = nonZeroBalances.map(async (token: AlchemyTokenBalance) => {
```

Comment:
Moralis gives us all the data in one single call. Definitely worth using their API as this can get very annoying with high amounts of tokens

https://docs.moralis.com/web3-data-api/evm/reference/wallet-api/get-token-balances-by-wallet?address=0xd8dA6BF26964aF9D7eEd9e03E53415D37aA96045&chain=base&token_addresses=[]